plugins/blockvolume/blockprovider/blockvolume_provider.go

```diff
@@ -0,0 +1,59 @@
+package blockprovider
+
+import (
+	"fmt"
+	"sync"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// ProviderFunc returns a block Provider instance. It also returns an error
+// If occurred any while creating a Provider instance
+type ProviderFunc func() (Provider, error)
+
+var (
+	providersMutex  sync.Mutex
+	providerFactory = make(map[string]ProviderFunc)
+)
+
+// Provider is an abstract, pluggable interface for block volume providers
+type Provider interface {
+	CreateBlockVolume(name string, size int64, hosts []string, options ...BlockVolOption) (BlockVolume, error)
+	DeleteBlockVolume(name string, options ...BlockVolOption) error
+	GetBlockVolume(id string) (BlockVolume, error)
+	BlockVolumes() []BlockVolume
+	ProviderName() string
+}
+
+// BlockVolume is an interface which provides information about a block volume
+type BlockVolume interface {
+	Name() string
+	HostVolume() string
+	HostAddresses() []string
+	IQN() string
+	Username() string
+	Password() string
+	Size() uint64
+}
+
+// RegisterBlockProvider will register a block provider
+func RegisterBlockProvider(name string, f ProviderFunc) {
+	providersMutex.Lock()
```

Comment:
can you explain whats the advantage of taking lock here and in below functions.

Reply:
just to make sure providerFactory map is not accessed parallelly by some other goroutine